### ISSUE TITLE
纯文本数据+添加 special token，全参数微调训练Qwen2VL，模型不收敛

### ISSUE THREAD

**Issue by AIFFFENG** [modelscope/ms-swift#3804]
**Describe the bug**
What the bug is, and how to reproduce, better with screenshots(描述bug以及复现过程，最好有截图)
尝试使用10000以内的加减法纯文本数据进行训练，这些数字均已使用special token替换。比如数字"1"替换为"<1>"，尽管这些special token在框架中验证有效，但在进行全参数微调训练Qwen2VL模型时，模型无法收敛。Token acc一直没有超过0.5

**Comment by AIFFFENG** [comment]
> **Describe the bug** What the bug is, and how to reproduce, better with screenshots(描述bug以及复现过程，最好有截图) 
尝试使用10000以内的加减法纯文本数据进行训练，这些数字均已使用special token替换。尽管这些special token在框架中验证有效，但在进行全参数微调训练Qwen2VL模型时，仍遇到了收敛性问题。Token acc一直没有超过0.5

**Comment by Jintao-Huang** [comment]
shell发一下看看

以及loss曲线

**Comment by AIFFFENG** [comment]
> shell发一下看看
> 
> 以及loss曲线

shell为
args="--deepspeed zero3 \
    --model /mnt/workspace/VLA/base_model/Qwen2-VL-2B-Instruct \
    --train_type full \
    --dataset '/mnt/workspace/VLA/dataset/Pretraining-0324-取50张图片-开始点/数字/forqwen2_5_train.jsonl' \
    --torch_dtype bfloat16 \
    --freeze_aligner false \
    --num_train_epochs 10 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --lazy_tokenize true \
    --learning_rate 5e-5 \
    --split_dataset_ratio 0.0 \
    --dataset_num_proc 32 \
    --truncation_strategy delete \
    --gradient_accumulation_steps 1 \
    --freeze_vit false \
    --save_steps 200 \
    --logging_steps 5 \
    --max_length 30000 \
    --lr_scheduler_type cosine \
    --output_dir /mnt/workspace/VLA/output_model/0408_just_number \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 8 \
    --add_version \
    --attn_impl flash_attn"

loss截取一部分
{"loss": 5.07161598, "token_acc": 0.38461538, "grad_norm": 6.5291843, "learning_rate": 4.893e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177161, "epoch": 1.38731267, "global_step/max_steps": "9535/68730", "percentage": "13.87%", "elapsed_time": "14h 56m 58s", "remaining_time": "3d 20h 48m 34s"}
{"loss": 4.73901863, "token_acc": 0.5, "grad_norm": 7.22556588, "learning_rate": 4.893e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177181, "epoch": 1.38804016, "global_step/max_steps": "9540/68730", "percentage": "13.88%", "elapsed_time": "14h 57m 20s", "remaining_time": "3d 20h 47m 26s"}
{"loss": 5.01708527, "token_acc": 0.5, "grad_norm": 8.94322703, "learning_rate": 4.893e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177202, "epoch": 1.38876764, "global_step/max_steps": "9545/68730", "percentage": "13.89%", "elapsed_time": "14h 57m 42s", "remaining_time": "3d 20h 46m 18s"}
{"loss": 4.87478027, "token_acc": 0.5, "grad_norm": 6.5805787, "learning_rate": 4.893e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177226, "epoch": 1.38949513, "global_step/max_steps": "9550/68730", "percentage": "13.89%", "elapsed_time": "14h 58m 3s", "remaining_time": "3d 20h 45m 7s"}
{"loss": 4.91479759, "token_acc": 0.5, "grad_norm": 6.61788897, "learning_rate": 4.892e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177249, "epoch": 1.39022261, "global_step/max_steps": "9555/68730", "percentage": "13.90%", "elapsed_time": "14h 58m 24s", "remaining_time": "3d 20h 43m 55s"}
{"loss": 4.88957291, "token_acc": 0.5, "grad_norm": 8.08351807, "learning_rate": 4.892e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177271, "epoch": 1.39095009, "global_step/max_steps": "9560/68730", "percentage": "13.91%", "elapsed_time": "14h 58m 45s", "remaining_time": "3d 20h 42m 45s"}
{"loss": 4.94309082, "token_acc": 0.5, "grad_norm": 14.22372204, "learning_rate": 4.892e-05, "memory(GiB)": 9.64, "train_speed(iter/s)": 0.177295, "epoch": 1.39167758, "global_step/max_steps": "9565/68730", "percentage": "13.92%", "elapsed_time": "14h 59m 6s", "remaining_time": "3d 20h 41m 31s"}

**Comment by AIFFFENG** [comment]
> shell发一下看看
> 
> 以及loss曲线

![Image](https://github.com/user-attachments/assets/9ded6b96-6859-4bc5-9db2-48c5e9b0f8df)
![Image](https://github.com/user-attachments/assets/0610b609-82bc-4801-bdf8-879bbfb40c23)

**Comment by AIFFFENG** [comment]
@Jintao-Huang 有空的话帮忙看一下呗，多谢了

**Comment by Jintao-Huang** [comment]
没看出原因诶

**Comment by AIFFFENG** [comment]
> 没看出原因诶

但是我感觉加减法这个任务，模型应该能够很快收敛呀。我其实尝试了不用特殊token执行这个任务，模型收敛的巨快，损失直接接近0了 。而现在反而收敛不了，会不会是因为纯文本任务训练Qwen2VL这种多模态大模型，没有训练embed_token层？@Jintao-Huang

**Comment by Jintao-Huang** [comment]
我使用这个shell训练im_end token是正常的
```
nproc_per_node=2

CUDA_VISIBLE_DEVICES=0,1 \
NPROC_PER_NODE=$nproc_per_node \
swift sft \
    --model Qwen/Qwen2-VL-2B \
    --train_type full \
    --dataset 'AI-ModelScope/alpaca-gpt4-data-zh#500' \
              'AI-ModelScope/alpaca-gpt4-data-en#500' \
              'swift/self-cognition' \
    --torch_dtype bfloat16 \
    --num_train_epochs 10 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-5 \
    --gradient_accumulation_steps $(expr 16 / $nproc_per_node) \
    --eval_steps 200 \
    --save_steps 200 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --max_length 2048 \
    --output_dir output \
    --system 'You are a helpful assistant.' \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --model_author swift \
    --model_name swift-robot \
    --deepspeed zero2

```

**Comment by AIFFFENG** [comment]
> 我使用这个shell训练im_end token是正常的
> 
> ```
> nproc_per_node=2
> 
> CUDA_VISIBLE_DEVICES=0,1 \
> NPROC_PER_NODE=$nproc_per_node \
> swift sft \
>     --model Qwen/Qwen2-VL-2B \
>     --train_type full \
>     --dataset 'AI-ModelScope/alpaca-gpt4-data-zh#500' \
>               'AI-ModelScope/alpaca-gpt4-data-en#500' \
>               'swift/self-cognition' \
>     --torch_dtype bfloat16 \
>     --num_train_epochs 10 \
>     --per_device_train_batch_size 1 \
>     --per_device_eval_batch_size 1 \
>     --learning_rate 1e-5 \
>     --gradient_accumulation_steps $(expr 16 / $nproc_per_node) \
>     --eval_steps 200 \
>     --save_steps 200 \
>     --save_total_limit 2 \
>     --logging_steps 5 \
>     --max_length 2048 \
>     --output_dir output \
>     --system 'You are a helpful assistant.' \
>     --warmup_ratio 0.05 \
>     --dataloader_num_workers 4 \
>     --model_author swift \
>     --model_name swift-robot \
>     --deepspeed zero2
> ```

我下午又做了一些实验（3以内的加法，只给了9条数据，<0>+<0> <0>+<1> <0>+<2>  <1>+<1>...，发现模型确实能够收敛，但是收敛的很慢，训练了10个epoch的训练集acc也只有90）按理说不应该吧，这还是全量微调。可以再帮忙看一下吗？

![Image](https://github.com/user-attachments/assets/d37e30c8-1277-421b-81e6-9900534f08f2)
![Image](https://github.com/user-attachments/assets/e186ab82-82a4-49ad-8c04-3a385c013206)

**Comment by AIFFFENG** [comment]
grad_norm也是巨大
![Image](https://github.com/user-attachments/assets/a69268c2-53fe-41ab-8bf5-fa0c2b8ed86e)

**Comment by AIFFFENG** [comment]
@Jintao-Huang 我试了下Qwen2.5纯语言模型，也出现了这个问题，会不会是添加token的方式不对？能提供下添加token的代码吗？

**Comment by zjc6666** [comment]
@AIFFFENG  我在训练的时候添加了token后，也遇到了相同的不收敛问题， 请问后来问题有解决吗？

**Comment by z0o0ey** [comment]
please retry with latest swift and vllm, feel free to create a new issue if it the problem persists.
